### PR TITLE
PermutedIndex helper function

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #2307 

---

# Add helper functions to PermutedIndex needed for shuffling

includes:
PermutedIndex

**Write why you are making the changes in this pull request**
to comply with 0.6 spec

**Write a summary of the changes you are making**
changes in:
- beacon-chain/utils/shuffle_test.go
- beacon-chain/utils/shuffle.go
**Link anything that would be helpful or relevant to the reviewers**
https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#get_permuted_index

### CODE REVIEW

File: beacon-chain/utils/shuffle.go [hunk 1]
```diff
@@ -70,3 +79,160 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+	}
+	return index, nil
+}
+
+// BytesToInt returns the uint64 representation of a byte slice.
+//
+// Spec pseudocode definition:
+// def bytes_to_int(data: bytes) -> int:
+//     return int.from_bytes(data, 'little')
+func BytesToInt(data []byte) uint64 {
```
Comment (nisdas) [inline beacon-chain/utils/shuffle.go]: we already have all these helpers https://github.com/prysmaticlabs/prysm/blob/master/shared/bytesutil/bytes.go

Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: removing, thanks

Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: removed

File: beacon-chain/utils/shuffle.go [hunk 2]
```diff
@@ -3,12 +3,21 @@ package utils
...
+	"fmt"
+	"math"
 
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
 	"github.com/prysmaticlabs/prysm/shared/hashutil"
 	"github.com/prysmaticlabs/prysm/shared/params"
 )
 
+const hSeedSize = int8(32)
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Please remove the `h` prefix. It is not clear what that means.

File: beacon-chain/utils/shuffle.go [hunk 3]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+//         source = hash(seed + int_to_bytes1(round) + int_to_bytes4(position // 256))
+//         byte = source[(position % 256) // 8]
+//         bit = (byte >> (position % 8)) % 2
+//         index = flip if bit else index
+//     return index
+func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error) {
+	if params.BeaconConfig().ShuffleRoundCount == 0 {
+		return index, nil
+	}
+	// spec: assert index < list_size
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Remove this comment, its not necessary since the spec is in the godoc

File: beacon-chain/utils/shuffle.go [hunk 4]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+	if index >= listSize {
+		return 0, fmt.Errorf("input index %d out of bounds: %d",
+			index, listSize)
+	}
+	// spec: assert list_size <= 2**40
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Remove this comment, its not necessary since the spec is in the godoc

File: beacon-chain/utils/shuffle.go [hunk 5]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+	buf := make([]byte, hTotalSize, hTotalSize)
+	// Seed is always the first 32 bytes of the hash input, we never have to change this part of the buffer.
+	copy(buf[:32], seed[:])
+	for round := uint8(0); round < uint8(params.BeaconConfig().ShuffleRoundCount); round++ {
+		buf[hSeedSize] = round
+		hash := hashutil.Hash(buf[:hPivotViewSize])
+		hash8 := hash[:8]
+		pivot := bytesutil.FromBytes8(hash8) % listSize
+		flip := (pivot - index) % listSize
+		// spec: position = max(index, flip)
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Remove this comment, its not necessary since the spec is in the godoc

File: beacon-chain/utils/shuffle.go [hunk 6]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		// Why? Don't do double work: we consider every pair only once.
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Please use a better comment here. No need to ask the reader "why?"

Suggestion:
```go
// Consider every pair only once by picking the highest pair index to retrieve randomness.
```

File: beacon-chain/utils/shuffle.go [hunk 7]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		// (Otherwise we would swap it back in place)
+		// Pick the highest index of the pair as position to retrieve randomness with.
+		position := index
+		if flip > position {
+			position = flip
+		}
+		// spec: source = hash(seed + int_to_bytes1(round) + int_to_bytes4(position // 256))
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Remove this comment, its not necessary since the spec is in the godoc

File: beacon-chain/utils/shuffle.go [hunk 8]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		// - seed is still in 0:32 (excl., 32 bytes)
+		// - round number is still in 32
+		// - mix in the position for randomness, except the last byte of it,
+		//     which will be used later to select a bit from the resulting hash.
+		p4b := bytesutil.Bytes4(position >> 8)
+		copy(buf[hPivotViewSize:], p4b[:])
+		source := hashutil.Hash(buf)
+		// spec: byte = source[(position % 256) // 8]
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Remove this comment, its not necessary since the spec is in the godoc

File: beacon-chain/utils/shuffle.go [hunk 9]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		position := index
+		if flip > position {
+			position = flip
+		}
+		// spec: source = hash(seed + int_to_bytes1(round) + int_to_bytes4(position // 256))
+		// - seed is still in 0:32 (excl., 32 bytes)
+		// - round number is still in 32
+		// - mix in the position for randomness, except the last byte of it,
+		//     which will be used later to select a bit from the resulting hash.
+		p4b := bytesutil.Bytes4(position >> 8)
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: please choose a better name than p4b. It is impossible for the reader to understand what that means.

File: beacon-chain/utils/shuffle.go [hunk 10]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		// spec: position = max(index, flip)
+		// Why? Don't do double work: we consider every pair only once.
+		// (Otherwise we would swap it back in place)
+		// Pick the highest index of the pair as position to retrieve randomness with.
+		position := index
+		if flip > position {
+			position = flip
+		}
+		// spec: source = hash(seed + int_to_bytes1(round) + int_to_bytes4(position // 256))
+		// - seed is still in 0:32 (excl., 32 bytes)
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Hmm, this is hard to follow/understand. 
The seed is in bytes 0 through 31, the round number is at index 32.

File: beacon-chain/utils/shuffle.go [hunk 11]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		// - round number is still in 32
+		// - mix in the position for randomness, except the last byte of it,
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: What does this sentence mean?

File: beacon-chain/utils/shuffle.go [hunk 12]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		//     which will be used later to select a bit from the resulting hash.
+		p4b := bytesutil.Bytes4(position >> 8)
+		copy(buf[hPivotViewSize:], p4b[:])
+		source := hashutil.Hash(buf)
+		// spec: byte = source[(position % 256) // 8]
+		// Effectively keep the first 5 bits of the byte value of the position,
+		//  and use it to retrieve one of the 32 (= 2^5) bytes of the hash.
+		byteV := source[(position&0xff)>>3]
+		// Using the last 3 bits of the position-byte, determine which bit to get from the hash-byte (8 bits, = 2^3)
+		// spec: bit = (byte >> (position % 8)) % 2
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Remove this comment, its not necessary since the spec is in the godoc

File: beacon-chain/utils/shuffle.go [hunk 13]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		bitV := (byteV >> (position & 0x7)) & 0x1
+		//index = flip if bit else index
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: ??

File: beacon-chain/utils/shuffle.go [hunk 14]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+	if params.BeaconConfig().ShuffleRoundCount == 0 {
+		return index, nil
+	}
+	// spec: assert index < list_size
+	if index >= listSize {
+		return 0, fmt.Errorf("input index %d out of bounds: %d",
+			index, listSize)
+	}
+	// spec: assert list_size <= 2**40
+	if listSize > uint64(math.Pow(2, 40)) {
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Please make this a constant, `maxShuffleListSize`

File: beacon-chain/utils/shuffle.go [hunk 15]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		return 0, fmt.Errorf("list size %d out of bounds",
+			listSize)
+	}
+	buf := make([]byte, hTotalSize, hTotalSize)
+	// Seed is always the first 32 bytes of the hash input, we never have to change this part of the buffer.
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Please add comments to the godoc as well about the memory mapping and allocation for this implementation

File: beacon-chain/utils/shuffle.go [hunk 16]
```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
 	}
 	return divided
 }
+
+// PermutedIndex Return `p(index)` in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
+//     Utilizes 'swap or not' shuffling
+//
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Please also add somewhere in this godoc that this implementation is based on the original implementation from protolambda, https://github.com/protolambda/eth2-shuffle

File: beacon-chain/utils/shuffle.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: thanks, done.

File: beacon-chain/utils/shuffle.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: removed

File: beacon-chain/utils/shuffle.go (hunk 11 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: tried to make it more coherent 
```// add position except its last byte to []buf for randomness, ,
// it will be used later to select a bit from the resulting hash.```

File: beacon-chain/utils/shuffle.go (hunk 12 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: done

File: beacon-chain/utils/shuffle.go (hunk 16 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: important. done

File: beacon-chain/utils/shuffle.go (hunk 15 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: added

File: beacon-chain/utils/shuffle.go (hunk 14 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: done

File: beacon-chain/utils/shuffle.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: replaced

File: beacon-chain/utils/shuffle.go (hunk 13 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: it implements the spec in a little different way. practically doing the same without math operations

**Comment by shayzluf** [comment]
> Why delete `beacon-chain/chaintest/tests/shuffle-tests/shuffle.yaml `, can't we just skip the test? Dont think we should delete because it'll be lost

I have added the updated test yaml from here https://github.com/ethereum/eth2.0-tests/tree/master/shuffling

**Comment by shayzluf** [comment]
and updated the test struct to support it

File: beacon-chain/utils/shuffle.go [hunk 17]
```diff
@@ -70,3 +79,74 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
 	}
 	return divided
 }
+
+// PermutedIndex Return `p(index)` in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
+//     Utilizes 'swap or not' shuffling
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]: can you format the comments?

File: beacon-chain/utils/shuffle.go [hunk 18]
```diff
@@ -70,3 +79,74 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		if flip > position {
+			position = flip
+		}
+		// add position except its last byte to []buf for randomness,
+		// it will be used later to select a bit from the resulting hash.
+		p4b := bytesutil.Bytes4(position >> 8)
+		copy(buf[pivotViewSize:], p4b[:])
+		source := hashutil.Hash(buf)
+		// Effectively keep the first 5 bits of the byte value of the position,
+		//  and use it to retrieve one of the 32 (= 2^5) bytes of the hash.
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]: format the comments

File: beacon-chain/utils/shuffle_test.go [hunk 19]
```diff
@@ -72,3 +73,17 @@ func TestSplitIndices_OK(t *testing.T) {
 		}
 	}
 }
+
+func BenchmarkPermuteIndex(b *testing.B) {
+	listSizes := []uint64{4000000, 40000, 400}
+	// "random" seed for testing. Can be any 32 bytes.
```
Comment (terencechain) [inline beacon-chain/utils/shuffle_test.go]:
```suggestion
	// Random 32 bytes seed for testing.
```

File: beacon-chain/utils/shuffle_test.go [hunk 20]
```diff
@@ -72,3 +73,17 @@ func TestSplitIndices_OK(t *testing.T) {
...
+	seed := [32]byte{123, 42}
+	for _, listSize := range listSizes {
+		// benchmark!
```
Comment (terencechain) [inline beacon-chain/utils/shuffle_test.go]:
```suggestion
```

File: beacon-chain/chaintest/backend/shuffle_test_format.go [hunk 21]
```diff
@@ -12,7 +12,39 @@ type ShuffleTest struct {
...
+}
+
+// Validators --
+type Validators struct {
+	ActivationEpoch uint64 `yaml:"activation_epoch"`
+	ExitEpoch       uint64 `yaml:"exit_epoch"`
+	OriginalIndex   int    `yaml:"original_index"`
+}
+
+// type AutoGenerated struct {
```
Comment (prestonvanloon) [inline beacon-chain/chaintest/backend/shuffle_test_format.go]: Why is this commented out?

File: beacon-chain/utils/shuffle.go (hunk 17 above)
Comment (prestonvanloon) [reply beacon-chain/utils/shuffle.go]: +1

File: beacon-chain/utils/shuffle.go [hunk 22]
```diff
@@ -70,3 +79,74 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
 	}
 	return divided
 }
+
+// PermutedIndex Return `p(index)` in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]:
```suggestion
// PermutedIndex returns `p(index)` in a pseudorandom permutation `p` of `0...list_size - 1` with `seed` as entropy.
```

File: beacon-chain/utils/shuffle.go (hunk 17 above)
Comment (prestonvanloon) [reply beacon-chain/utils/shuffle.go]: Please use godocs locally to ensure these are formatted nicely

File: beacon-chain/utils/shuffle.go [hunk 23]
```diff
@@ -70,3 +79,74 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		hash8 := hash[:8]
+		pivot := bytesutil.FromBytes8(hash8) % listSize
+		flip := (pivot - index) % listSize
+		// spec: position = max(index, flip)
+		// Consider every pair only once by picking the highest pair index to retrieve randomness.
+		position := index
+		if flip > position {
+			position = flip
+		}
+		// add position except its last byte to []buf for randomness,
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]:
```suggestion
		// Add position except its last byte to []buf for randomness,
```

File: beacon-chain/utils/shuffle.go (hunk 18 above)
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]:
```suggestion
		// and use it to retrieve one of the 32 (= 2^5) bytes of the hash.
```

File: beacon-chain/utils/shuffle.go [hunk 24]
```diff
@@ -70,3 +79,74 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		// it will be used later to select a bit from the resulting hash.
+		p4b := bytesutil.Bytes4(position >> 8)
+		copy(buf[pivotViewSize:], p4b[:])
+		source := hashutil.Hash(buf)
+		// Effectively keep the first 5 bits of the byte value of the position,
+		//  and use it to retrieve one of the 32 (= 2^5) bytes of the hash.
+		byteV := source[(position&0xff)>>3]
+		// Using the last 3 bits of the position-byte, determine which bit to get from the hash-byte (8 bits, = 2^3)
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]:
```suggestion
		// Using the last 3 bits of the position-byte, determine which bit to get from the hash-byte (note: 8 bits = 2^3).
```

File: beacon-chain/utils/shuffle.go [hunk 25]
```diff
@@ -70,3 +79,74 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
 	}
 	return divided
 }
+
+// PermutedIndex Return `p(index)` in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
+//     Utilizes 'swap or not' shuffling
+// in this implementation we are allocating the memory with the seed that stays constant between iterations instead
+// of reallocating it each iteration as in spec.
+//	this implementation is based on the original implementation from protolambda, https://github.com/protolambda/eth2-shuffle
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]:
```suggestion
// This implementation is based on the original implementation from protolambda, https://github.com/protolambda/eth2-shuffle.
```

File: beacon-chain/utils/shuffle.go (hunk 9 above)
Comment (terencechain) [reply beacon-chain/utils/shuffle.go]: better name than p4b?

**Comment by prestonvanloon** [comment]
I'm still hopeful for a better name than `p4b`, but I don't feel that strongly about it. If you look at the code, the reader can see that it is position as 4 bytes.

**Comment by nisdas** [comment]
@prestonvanloon @terencechain Changed the naming of the variable to be more clear
